platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/EmbeddedSolrFiles.java

```diff
@@ -0,0 +1,178 @@
+   *     schemaXmls</code>, or <code>configProxy</code> is <code>null</code> or if unable to find
+   *     any files
+   */
+  public EmbeddedSolrFiles(
+      String coreName, String configXml, String[] schemaXmls, ConfigurationFileProxy configProxy) {
+    Validate.notNull(coreName, "invalid null Solr core name");
+    Validate.notNull(configXml, "invalid null Solr configuration file");
+    Validate.notNull(schemaXmls, "Invalid null Solr schema files");
+    Validate.notEmpty(schemaXmls, "missing Solr schema files");
+    Validate.noNullElements(schemaXmls, "invalid null Solr schema file");
+    Validate.notNull(configProxy, "invalid null Solr config proxy");
+    this.coreName = coreName;
+    this.configName = configXml;
+    this.configFile = EmbeddedSolrFactory.getConfigFile(configXml, configProxy, coreName);
+    Validate.notNull(configFile, "Unable to find Solr configuration file");
+    File solrSchemaFile = null;
+    String schemaXml = null;
+
+    for (final String s : schemaXmls) {
+      schemaXml = s;
+      solrSchemaFile = EmbeddedSolrFactory.getConfigFile(schemaXml, configProxy, coreName);
+      if (solrSchemaFile != null) {
+        break;
+      }
+    }
+    Validate.notNull(solrSchemaFile, "Unable to find Solr schema file.");
+    this.schemaFile = solrSchemaFile;
+    this.schemaName = schemaXml;
+  }
+
+  /**
+   * Gets the home directory where the configuration resides.
+   *
+   * @return the home directory where the configuration resides
+   */
+  public File getConfigHome() {
+    return configFile.getParentFile();
+  }
+
+  /**
+   * Creates or retrieves the schma index corresponding to this set of Solr files.
+   *
+   * @return retrieves a corresponding index for the schema
+   * @throws IllegalArgumentException if unable to load or parse the corresponding schema or config
+   */
+  public IndexSchema getSchemaIndex() {
+    if (schemaIndex == null) {
+      final SolrConfig cfg = getConfig(); // make sure it is initialized
+
+      LOGGER.debug(
+          "Loading and creating index for {} schema using file [{} ({})]",
+          coreName,
+          schemaName,
+          schemaFile);
+      try {
+        this.schemaIndex =
+            new IndexSchema(
+                cfg, schemaName, new InputSource(FileUtils.openInputStream(schemaFile)));
+      } catch (IOException e) {
+        LOGGER.debug(
+            "failed to parse {} Solr schema file [{} ({})]", coreName, schemaName, schemaFile, e);
+        throw new IllegalArgumentException("Unable to parse Solr schema file: " + schemaName, e);
+      }
+    }
+    return schemaIndex;
+  }
+
+  /**
+   * Creates or retrieves the Solr config for this set of Solr files.
+   *
+   * @return retrieves a corresponding Solr config
+   * @throws IllegalArgumentException if unable to load or parse the corresponding config
+   */
+  public SolrConfig getConfig() {
+    if (solrConfig == null) {
```

Comment:
❓Does this code need to be thread-safe? `solrConfig` is `volatile`, but that won't prevent two threads from executing this method at the same time. May not matter since you'd expect all the threads to set the variable to the same thing, but I thought I'd ask :-)

Reply:
Shouldn't matter. It will only get done twice in a race condition.